Set up server files and create user register endpoint #2

enricodelarosa opened the pull request.

idiglove commented on controllers/utils/validate.js:
```diff
@@ -0,0 +1,56 @@
+    const isValid= validate.noneIsBlank(reqBody,neededKeys) && validate.countOfKeysMatch(reqBody, neededKeys);
+
+    return {reqBody, isValid}
+}
```
im seeing a lot of misspellings and it should  be `isValid = `

idiglove commented on routes/api/apiRoutes.js:
```diff
@@ -0,0 +1,14 @@
+
+apiRouter.use('/users', usersRouter);
+
+apiRouter.use('/games', gamesRouter); 
```
shouldnt we have 

apiRouter.use('/reviews', reviewsRouter);

enricodelarosa replied:
Oh right, we need a get all reviews endpoint. will and it by tomorrow

idiglove reviewed (commented):
just a few things, and as discussed let's add TS support on it 👌🏽

enricodelarosa commented:
Already contains prisma schema for user, user_role, game, and review

idiglove commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,48 @@
+  username   String    @unique(map: "username") @db.VarChar(50)
+  first_name String    @db.VarChar(50)
+  last_name  String    @db.VarChar(50)
+  password   String    @db.Char(60)
+  role       String    @default("user") @db.VarChar(80)
+  user_role  user_role @relation(fields: [role], references: [id], onUpdate: Restrict, map: "user_ibfk_1")
```
what is the diff of role and user_role?

enricodelarosa replied:
I'm actually not sure. What I did was create the user and user_role tables on mysql cli and use prisma db pull to generate the prisma schema. Will look into what line 18 means. But right now I think it establishes a one to many relation with the user_role table. Will look into what map means.

enricodelarosa replied:
I looked it up in the prisma docs and i got.

onUpdate: Restrict - prevents the identifier of a referenced record from being changed.
In the context of the user_role and user tables, this means that user_roles id cannot be changed.

i can't find map: in the doccs, let me try to remove it and see what happens

enricodelarosa replied:
map: was using the user_ibkf_1 index for some reason. I suspect it's something that present on my tables since I created the tables initially using mysql cli and then did prisma db pull. I removed map in that relation and just used db push to make the tables according to the prisma schema.

idiglove replied:
Hi Rico, I don't think we have to worry about the map. How does the fields `role` and `user_role` relate to each other? I'm trying to understand hehe

enricodelarosa replied:
Hi Fath, yes I did remove the map. A user can only have a role that is present in the user_role table. is that okay?

idiglove replied:
What is the use of user_role table?

enricodelarosa replied:
I wanted to set the role column of the user table to be a foreign key referencing the id column of the user_role table to establish a constraint. This would prevent the back end code from adding users with roles not present in the table. But now that I think about it, user_role table could actually be removed. Should I just remove user_role table?

roles available would be user, admin, superadmin and more can be added if necessary

idiglove commented on controllers/utils/validate.ts:
```diff
@@ -0,0 +1,37 @@
+
+export const email = (email: any) => {
+    return emailValidator.validate(email);
+}
+
+export const noneIsBlank = (Obj: any, neededBodyKeys: any) => {
+
+    for(let key of neededBodyKeys) {
+        if (String(Obj[key]).length == 0) {
+            return false;
+        }
+    }
+    return true;
+}
+
+export const countOfKeysMatch = (Obj: any, neededKeys: any) => {
+
+    const countOfKeys = Object.keys(Obj).length;
+    return neededKeys.length <= countOfKeys; 
+
```
can we make sure that we don't use the type: `any` ?
Also, can we add a prettier/linter so that we don't have unnecessary whitespace and linebreaks?

idiglove commented on controllers/usersController.ts:
```diff
@@ -0,0 +1,122 @@
+    // I want to make this to a middleware? Or like a reusable function so that the isValid checking is not repeated and the reqBody properties would be available already as variables.
+    const neededKeys = ['email', 'username', 'firstName', 'lastName', 'password'];
+
+    const { reqBody, isValid } = validate.bodyVsKeys(req.body, neededKeys);
```
reading `validate.bodyVsKeys` doesnt make sense, can we rename it to something more understandable?

enricodelarosa replied:
Got it. Renaming now

idiglove commented on controllers/usersController.ts:
```diff
@@ -0,0 +1,122 @@
+export const registerUser = async (req: any, res: any) => {
+
+    // I want to make this to a middleware? Or like a reusable function so that the isValid checking is not repeated and the reqBody properties would be available already as variables.
+    const neededKeys = ['email', 'username', 'firstName', 'lastName', 'password'];
```
yes you can do that
https://expressjs.com/en/guide/writing-middleware.html

idiglove commented on controllers/utils/validate.ts:
```diff
@@ -0,0 +1,37 @@
+        return {reqBody, isValid: false}
+    }
+
+    // Is valid true only if noen of the requeried fields is blank and the count of properties of req.body object is less than equal to the nubmer of needed keys
+    const isValid = noneIsBlank(reqBody,neededKeys) && countOfKeysMatch(reqBody, neededKeys);
```
im seeing a lot of misspellings on the comments

idiglove replied:
like the engr manager from paymongo yesterday, he says that let's limit comments and instead use understandable code by making our function and variable names readable and easy to understand
in this case, I don't understand what noneIsBlank, and  countOfKeysMatch

enricodelarosa commented:
I rewrote `validate.ts` so that the default export can be used as middleware.

Also changed the variable names so it's more readable. You're right, implying what the code does by using the variable and function names is pretty elegant! I hope I did it right haha.

Installed prettier and set `tabWidth` to 4 spaces.

Removed `any` types and replaced with more specific ones.

enricodelarosa commented:
Added `outDir: dist` to tsconfig.json file.

enricodelarosa commented:
Renamed files for consistency also added `"watch": true` to tsconfig.json

enricodelarosa commented:
Renamed files for consistency also added `"watch": true` to tsconfig.json

EstopaceMA reviewed (changes requested):
I think you should add a dist folder to be ignored in the gitignore file as it's not necessary to be committed since they're just locally generated build files.

enricodelarosa commented:
Added dist folder to `.gitignore` file

idiglove commented on controllers/utils/validate.ts:
```diff
@@ -0,0 +1,90 @@
+    const isReqBodyEmpty = Object.keys(reqBody).length === 0;
+    if (isReqBodyEmpty) {
+        objHasProps = false;
+        let returnObj: returnObj = {
```
im loving the names :)
although i think we should refrain from using same variable names as with typenames. And types/interfaces convention initial char starts as capital

enricodelarosa replied:
Got it will be addressing this

idiglove commented on controllers/utils/validate.ts:
```diff
@@ -0,0 +1,90 @@
+): returnObj => {
+    let objHasProps: boolean = true;
+    let missing: string[] = [];
+    let blank: string[] = [];
```
hmm what is blank?

enricodelarosa replied:
the props. It's defined (not missing) but has a blank value. Will change the property name to communicate that :)

idiglove commented on package.json:
```diff
@@ -0,0 +1,42 @@
+{
+    "name": "okunokumo.local",
```
can we rename to okunokumo.be instead?

enricodelarosa replied:
Got it

idiglove reviewed (approved):
💯

EstopaceMA reviewed (approved):
LGTM! 💯
